addons/image/bmp.c

```diff
@@ -297,387 +389,477 @@ static void read_4bit_line(int length, ALLEGRO_FILE *f, unsigned char *buf)
+      return false;
+   }
 
-         case 32:
-            read_32bit_line(infoheader->biWidth, f, data);
-            break;
-      }
-      if (infoheader->biBitCount <= 8) {
-         for (j = 0; j < (int)infoheader->biWidth; j++) {
-            if (keep_index) {
-               data[0] = buf[j];
-               data++;
+   decode_bitfield(infoheader->biRedMask, &rs, &rm);
+   decode_bitfield(infoheader->biGreenMask, &gs, &gm);
+   decode_bitfield(infoheader->biBlueMask, &bs, &bm);
+   decode_bitfield(infoheader->biAlphaMask, &as, &am);
+
+   for (i = 0; i < (int)(sizeof(tempconvert) / sizeof(int *)); ++i) {
+      uint32_t mask = ~(0xFFFFFFFFU << (i+1)) & 0xFFFFFFFFU;
+
+      switch (i) {
+         case 0: tempconvert[i] = _al_rgb_scale_1; break;
+         case 3: tempconvert[i] = _al_rgb_scale_4; break;
+         case 4: tempconvert[i] = _al_rgb_scale_5; break;
+         case 5: tempconvert[i] = _al_rgb_scale_6; break;
+         default:
+            if (rm == mask || gm == mask || bm == mask || am == mask) {
+               int entries = (1 << (i+1));
+
+               // Skip generating tables for tiny images
+               if (width * height > entries * 2)
+               {
+                  tempconvert[i] = al_malloc(sizeof(int) * entries);
+                  generate_scale_table(tempconvert[i], entries);
+               }
             }
             else {
-               data[0] = pal[buf[j]].r;
-               data[1] = pal[buf[j]].g;
-               data[2] = pal[buf[j]].b;
-               data[3] = 255;
-               data += 4;
+               tempconvert[i] = NULL;
             }
-         }
       }
+
+      if (rm == mask) rtable = tempconvert[i];
+      if (gm == mask) gtable = tempconvert[i];
+      if (bm == mask) btable = tempconvert[i];
+      if (am == mask) atable = tempconvert[i];
    }
 
-   al_free(buf);
-}
+   line = height < 0 ? 0 : height - 1;
+   dir = height < 0 ? 1 : -1;
+   height = abs(height);
 
+   for (i = 0; i < height; i++, line += dir) {
+      unsigned char *data = (unsigned char *)lr->data + lr->pitch * line;
 
+      bytes_read = al_fread(f, linebuf, linesize);
+      memset(linebuf + bytes_read, 0, linesize - bytes_read);
 
-/* read_RGB_alpha_image:
- *  For reading the non-compressed BMP image format (32-bit with alpha)
- *  Return false if the alpha mask was invalid
- */
-static bool read_RGB_alpha_image(ALLEGRO_FILE *f, int flags,
-   const BMPINFOHEADER *infoheader, ALLEGRO_LOCKED_REGION *lr)
-{
-   int i, line, height, dir;
-   unsigned char *buf;
-   unsigned char *data;
+      for (k = 0; k < width; k++) {
+         uint32_t pixel = read_32le(linebuf + k*bytes_per_pixel);
+         uint32_t r, g, b, a = 255;
 
-   int as, am;
+         r = ((pixel >> rs) & rm);
+         g = ((pixel >> gs) & gm);
+         b = ((pixel >> bs) & bm);
 
-   if (!decode_bitfield(infoheader->biAlphaMask, &as, &am))
-      return false;
+              if (rtable) r = rtable[r];
```

Comment:
This is really funky looking :P. If it's for alignment, could you align it by adding spaces after the `if` instead?